src/macaron/repo_finder/provenance_finder.py

```diff
@@ -35,191 +37,350 @@ def __init__(self) -> None:
                 elif isinstance(registry, JFrogMavenRegistry):
                     self.jfrog_registry = registry
 
-    def find_provenance(self, purl: PackageURL) -> InTotoPayload | None:
-        """Find the provenance files of the passed PURL.
+    def find_provenance(
+        self,
+        purl: PackageURL,
+        discovery_functions: list[Callable[..., list[InTotoPayload]]] | None = None,
+        parameters: list[Any] | None = None,
+    ) -> list[InTotoPayload]:
+        """Find the provenance file(s) of the passed PURL.
 
         Parameters
         ----------
         purl: PackageURL
             The PURL to find provenance for.
+        discovery_functions: list[Callable[..., list[InTotoPayload]]] | None
+            A list of discovery functions to use for the given PURL, or None if the default should be used instead.
+        parameters: list[Any] | None
+            The list of parameters to pass to the callable, or None.
 
         Returns
         -------
-        InTotoPayload | None
-            The provenance payload if found, or None.
+        list[InTotoPayload]
+            The provenance payload, or an empty list if not found.
         """
-        if determine_abstract_purl_type(purl) == AbstractPurlType.REPOSITORY:
-            # Do not perform this function for repository type targets.
-            return None
-
-        if purl.type == "npm":
-            if self.npm_registry:
-                return ProvenanceFinder.find_npm_provenance(purl, self.npm_registry)
-            logger.debug("Missing npm registry to find provenance in.")
-        elif purl.type in ["gradle", "maven"]:
-            if self.jfrog_registry:
-                return ProvenanceFinder.find_gav_provenance(purl, self.jfrog_registry)
-            logger.debug("Missing JFrog registry to find provenance in.")
-        else:
-            logger.debug("Provenance finding not supported for PURL type: %s", purl.type)
-
-        return None
-
-    @staticmethod
-    def find_npm_provenance(purl: PackageURL, npm_registry: NPMRegistry) -> InTotoPayload | None:
-        """Find and download the NPM based provenance for the passed PURL.
+        if not discovery_functions:
+            if determine_abstract_purl_type(purl) == AbstractPurlType.REPOSITORY:
+                # Do not perform default discovery for repository type targets.
+                return []
```

Comment:
When we provide a list of `discovery_functions`, do we also need to handle repository type PURL, or we let the discovery functions decide whether the they want to support it or not?

Reply:
When provided, the intention is to let them decide what to do with different PURL types. However, from other work I've been doing, I'm not sure there is a possibility of utilising repository PURLs at this point in the analysis.